lib/translate/common.h

```diff
@@ -0,0 +1,24 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIB_TRANSLATOR_COMMON_H
+#define LIB_TRANSLATOR_COMMON_H
+
+static void (*debug_callback)(const char*);
```

Comment:
This will define a separate, unique global variable in every file that includes it. To have a single global, place the definition inside a `.cc` file, without the `static` qualifier that makes it internal to the compilation unit. The header file should only contain a *declaration*, which indicates to including code that the global exists in another compilation unit. In this case a declaration would use the same syntax as the definition, but with the `extern` qualifier.

However, in this particular case we should go a step further. Exposing the global allows any including code to write to it, but for readability's sake we should ensure that the only way to write to the global is with the provided setter function. To achieve that, omit the global's declaration from the header entirely, and define a helper function named something like `DebugLog` for use within this library's code which just forwards its arguments to the function pointer, if set. That will also save downstream code from having to check whether the pointer's set everywhere.

Reply:
Thank you for the thorough explanation! I think I understand much better how the debug callback should work now :)